core/src/main/scala/cats/Representable.scala

```diff
@@ -0,0 +1,133 @@
+
+  implicit class TabulateOps[A, R](f: R => A) {
+    def tabulate[F[_]](implicit R: Representable.Aux[F, R]): F[A] = R.tabulate(f)
+  }
+}
+
+object Representable extends RepresentableSyntax {
+  type Aux[F[_], R] = Representable[F] { type Representation = R }
+
+  def apply[F[_]](implicit ev: Representable[F]): Representable[F] = ev
+
+  /**
+   * Witness for: Id[A] <-> Unit => A
+   */
+  implicit val catsRepresentableForId: Representable.Aux[Id, Unit] = new Representable[Id] {
+    override type Representation = Unit
+    override val F: Functor[Id] = Functor[Id]
+
+    override def tabulate[A](f: Unit => A): Id[A] = f(())
+
+    override def index[A](f: Id[A]): Unit => A = (_: Unit) => f
+  }
+
+  /**
+   * Witness for: E => A <-> E => A
+   */
+  implicit def catsRepresentableForFunction1[E]: Representable.Aux[E => ?, E] = new Representable[E => ?] {
+    override type Representation = E
+    override val F: Functor[E => ?] = Functor[E => ?]
+
+    override def tabulate[A](f: E => A): E => A = f
+
+    override def index[A](f: E => A): E => A = f
+  }
+
+  implicit val catsFunctorForTuple2: Functor[λ[P => (P, P)]] = new Functor[λ[P => (P, P)]] {
+    override def map[A, B](fa: (A, A))(f: A => B): (B, B) = (f(fa._1), f(fa._2))
+  }
+
+  /**
+   * Witness for: (A, A) <-> Boolean => A
+   */
+  implicit val catsRepresentableForTuple2: Representable.Aux[λ[P => (P, P)], Boolean] = new Representable[λ[P => (P, P)]] {
+    override type Representation = Boolean
+    override val F: Functor[λ[P => (P, P)]] = Functor[λ[P => (P, P)]]
+
+    override def tabulate[A](f: Boolean => A): (A, A) = (f(true), f(false))
+
+    override def index[A](pair: (A, A)): Boolean => A = {
+      case true => pair._1
+      case false => pair._2
+    }
+  }
+
+  /**
+   * Witness for: Kleisli[M, E, A] <-> (E, R) => A
+   * if M is Representable
+   */
+  implicit def catsRepresentableForKleisli[M[_], R, E](
+    implicit R: Representable.Aux[M, R], FK: Functor[Kleisli[M, E, ?]]): Representable.Aux[Kleisli[M, E, ?], (E, R)] =
+    new Representable[Kleisli[M, E, ?]] {
+
+      override type Representation = (E, R)
+
+      override val F: Functor[Kleisli[M, E, ?]] = FK
+
+      def index[A](f: Kleisli[M, E, A]): Representation => A = {
+        case (e, r) => R.index(f.run(e))(r)
+      }
+
+      def tabulate[A](f: Representation => A): Kleisli[M, E, A] = {
+        def curry[X, Y, Z](f: (X, Y) => Z): X => Y => Z = x => y => f(x, y)
+
+        Kleisli[M, E, A](curry(Function.untupled(f)) andThen R.tabulate)
+      }
+    }
+
+  def representableIsBimonad[F[_], R](implicit R: Representable.Aux[F, R], M: Monoid[R]): Monad[F] = new Bimonad[F] with StackSafeMonad[F] {
+    override def pure[A](x: A): F[A] = R.tabulate(_ => x)
+
+    override def flatMap[A, B](fa: F[A])(f: A => F[B]): F[B] =
+      R.tabulate(a => R.index(f(R.index(fa)(a)))(a))
+
+    // TODO: Can we make this monad stack safe?
+    // override def tailRecM[A, B](a: A)(f: A => F[Either[A, B]]): F[B] = ???
```

Comment:
I think something like this works:
```scala
R.tabulate { r: R =>
  @annotation.tailrec
  def loop(a: A): B = 
    R.index(f(a))(r) match {
      case Right(b) => b
      case Left(a) => loop(a)
    }

  loop(a)
}
```

Reply:
Thanks! I hadn't got around to thinking this through. That solution works.